### PR TITLE
Fix breakage due to rust-lang/rust#61988

### PR DESCRIPTION
**Pull request by mikerite**
changelog: none

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -590,9 +587,16 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Loops {
...
         }
 
-        // check for while loops which conditions never change
-        if let ExprKind::While(ref cond, _, _) = expr.node {
-            check_infinite_loop(cx, cond, expr);
+        if_chain! {
+            if let ExprKind::Loop(block, _, LoopSource::While) = &expr.node;
+            if let Block { expr: Some(expr), .. } = &**block;
+            if let ExprKind::Match(cond, arms, MatchSource::WhileDesugar) = &expr.node;
+            if let ExprKind::DropTemps(cond) = &cond.node;
```
Comment (Manishearth) [inline clippy_lints/src/loops.rs]: This should be a `while_desugar` in `utils` like the `if_let_desugar` and `for_desugar` we already have

File: clippy_lints/src/loops.rs [hunk 2]
```diff
@@ -481,17 +481,14 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Loops {
...
-            _ => (),
-        }
+        if let ExprKind::Loop(ref block, _, _) = expr.node {
+            match never_loop_block(block, expr.hir_id) {
+                NeverLoopResult::AlwaysBreak => {
+                    span_lint(cx, NEVER_LOOP, expr.span, "this loop never actually loops")
+                },
+                NeverLoopResult::MayContinueMainLoop | NeverLoopResult::Otherwise => (),
+            }
+        };
```
Comment (lzutao) [inline clippy_lints/src/loops.rs]: Is this `;` needed?

File: clippy_lints/src/utils/author.rs [hunk 3]
```diff
@@ -714,6 +702,7 @@ fn loop_desugaring_name(des: hir::LoopSource) -> &'static str {
     match des {
         hir::LoopSource::ForLoop => "LoopSource::ForLoop",
         hir::LoopSource::Loop => "LoopSource::Loop",
+        hir::LoopSource::While => "LoopSource::WhileDesugar",
```
Comment (lzutao) [inline clippy_lints/src/utils/author.rs]: `=> "LoopSource::While",`

**Comment by Manishearth** [comment]
@bors r+

thanks!

**Comment by lzutao** [comment]
@Manishearth reapprove?

**Pull request merged**